diff -urN old/net/wireless/core.c patch/net/wireless/core.c
--- old/net/wireless/core.c	2014-11-14 15:09:35.000000000 +0800
+++ patch/net/wireless/core.c	2014-11-17 15:28:05.000000000 +0800
@@ -951,8 +951,12 @@
 		cfg80211_update_iface_num(rdev, wdev->iftype, -1);
 		if (rdev->scan_req && rdev->scan_req->wdev == wdev) {
 			if (WARN_ON(!rdev->scan_req->notified))
-				rdev->scan_req->aborted = true;
+				rdev->scan_req->aborted = true;			
+#if 1 //rtk: client mode 2nd Association
+			cfg80211_scan_done(rdev->scan_req, false);
+#else	
 			___cfg80211_scan_done(rdev, false);
+#endif
 		}
 
 		if (WARN_ON(rdev->sched_scan_req &&
