Index: linux-3.10.49/arch/mips/Kbuild.platforms
===================================================================
--- linux-3.10.49.orig/arch/mips/Kbuild.platforms
+++ linux-3.10.49/arch/mips/Kbuild.platforms
@@ -31,7 +31,8 @@ platforms += sni
 platforms += txx9
 platforms += vr41xx
 platforms += wrppmc
-platforms += realtek
+platforms += rtl8197f
+platforms += realtek
 
 # include the platform specific files
 include $(patsubst %, $(srctree)/arch/mips/%/Platform, $(platforms))
Index: linux-3.10.49/arch/mips/Kconfig
===================================================================
--- linux-3.10.49.orig/arch/mips/Kconfig
+++ linux-3.10.49/arch/mips/Kconfig
@@ -445,15 +445,26 @@ config RALINK
 	select HAVE_MACH_CLKDEV
 	select CLKDEV_LOOKUP
 
-config REALTEK
-        bool "Realtek mips based machines"
+config REALTEK
+        bool "Realtek 98C based machines"
         select CEVT_R4K
         select CSRC_R4K
         select DMA_NONCOHERENT
         select IRQ_CPU
         select SYS_HAS_CPU_MIPS32_R2
         select SYS_SUPPORTS_32BIT_KERNEL
-	select SYS_SUPPORTS_BIG_ENDIAN
+		select SYS_SUPPORTS_BIG_ENDIAN
+        select SYS_HAS_EARLY_PRINTK
+
+config REALTEK_97F
+        bool "Realtek 97F based machines"
+        select CEVT_R4K
+        select CSRC_R4K
+        select DMA_NONCOHERENT
+        select IRQ_CPU
+        select SYS_HAS_CPU_MIPS32_R2
+        select SYS_SUPPORTS_32BIT_KERNEL
+		select SYS_SUPPORTS_LITTLE_ENDIAN
         select SYS_HAS_EARLY_PRINTK
 
 
@@ -869,7 +880,8 @@ source "arch/mips/cavium-octeon/Kconfig"
 source "arch/mips/loongson/Kconfig"
 source "arch/mips/loongson1/Kconfig"
 source "arch/mips/netlogic/Kconfig"
-source "arch/mips/realtek/Kconfig"
+source "arch/mips/rtl8197f/Kconfig"
+source "arch/mips/realtek/Kconfig"
 
 
 endmenu
